### PR TITLE
For discussion: use isort for formatting import

### PR DESCRIPTION
**Pull request by ikkoham**
### Summary

Terra introduced black formatter in #6361. However, there is currently no formatting for imports. 
Developer's style differences regarding line breaks and ordering of imports can cause merge conflicts
as well as make code difficult to read.  `isort` is a formatter compatible with black. The development 
flow will be the same as when black was introduced, and there will be no major changes for developers.

### Details and comments

if this direction looks good, I will modify contibution guidelines, etc.

### References

- isort: https://github.com/PyCQA/isort

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -21,3 +21,13 @@ before-test = "pip install --only-binary=numpy,scipy numpy scipy"
...
 before-all = "yum install -y wget && {package}/tools/install_rust.sh"
 environment = 'PATH="$PATH:$HOME/.cargo/bin"'
+
+[tool.isort]
+multi_line_output = 3
+include_trailing_comma = true
+force_grid_wrap = 0
+use_parentheses = true
+ensure_newline_before_comments = true
+force_sort_within_sections = true
```
Comment (t-imamichi) [inline pyproject.toml:31]: How about using `profile = black`?
https://pycqa.github.io/isort/docs/configuration/profiles.html

Comment (ikkoham) [reply pyproject.toml]: Yes. Thanks. The only difference is `line_length`. `force_sort_within_sections` is for pylint.

Done in dda3319

**Comment by jakelishman** [comment]
I'll preface this by saying that I don't feel very strongly.  Personally, I don't remember ever having had any serious issue caused by import order, either as a merge conflict or by having difficulty reading them.  If I have had a merge conflict with them, it was so trivial to solve that I don't remember it.  On the other hand, I'm concerned about the on-boarding cost to developers by adding another tool that they have to learn and configure if it's not providing us with much benefit.

My vote would be against this, as the supposed benefits don't seem sufficient to justify the cost to new or external developers, but I'm aware that's somewhat subjective and others might feel more strongly.  I think I used to be more in favour of this sort of tool, but as I work with others, and particularly getting new contributors started, it's dropped in my estimations a bit.  I like code formatters in general, I just don't like a fragmented ecosystem of several of them.

I know Lev's thought about doing this before as well (@levbishop).

**Comment by ikkoham** [comment]
Thanks @jakelishman. 
I understand what you are saying. I think there is a trade-off between quality and on-boarding cost.

I also don't remember when, but there were different developers with different isort settings, and every time they updated, they alternated the changes just like ping-pong...
It might be a good idea for us not to enforce, i.e. not to introduce checking format, but only to put in a setting (pyproject.toml).

**Comment by levbishop** [comment]
I did explore some of these ideas a few months ago on my branch https://github.com/levbishop/qiskit-terra/tree/isort
If you look at the individual commits on that branch you can see different levels of how invasive changes to make:

1. The first commit adds options to only group the imports into sections:
```
 [tool.isort]
profile = "black"
line_length = 100
treat_all_comments_as_code = "True"
only_sections = "True"
```
and adds some comments with the `treat_all_comments_as_code` in order to avoid circular import errors. I think I like this approach a bit better than just excluding `__init__.py` as in your approach. And the second commit commit runs isort with those options

2. The next few commits add `from __future__ import annotations` to all files, and runs pyupgrade and black to enable the associated type annotation improvements with some manual cleanup. Overall I like the readability improvements from this, but I'm not sure if it's a good idea to commit to it before python 3.11 is released with official mandatory requirement of PEP 563 (it was already bumped from python 3.10).

1. The next few commits use absolufy to replace absolute imports, which I don't think we've discussed much, but if we're going to have all this churn from isort already, then now would be the time to bite the bullet and do it (I much prefer relative import).

1. The next couple commits turn off `only_sections` and activate sorting within sections. I feel this is not much extra value on top of the previous changes, but I don't feel strongly. 

1. The next couple commits add in some options from the google style guide, **which I think would be a mistake for our codebase**. Those options make sense if you are following the full [google style guide](https://google.github.io/styleguide/pyguide.html):

> ```python
> from sound.effects import echo
> ...
> echo.EchoFilter(input, output, delay=0.7, atten=4)
> ```
vs
> ```
> from sound.effects.echo import EchoFilter
> ...
> EchoFilter(input, output, delay=0.7, atten=4)
> ```

I actually rather like the google recommendation here, **but** we don't generally follow it in qiskit, and without it the google isort settings make the import sections way too verbose. I don't think the effort to switch over to the google style of importing modules only would come close to being worthwhile (unless maybe there were a tool that could do it in the manner of `black`, but I haven't seen one, and even then it's probably too disruptive).

My current thoughts would be that **if** we make a change like this, then to do all of steps 1-3, probably don't do 4, definitely don't do 5.

I think with 1-3 only then its low-impact on contributors to add CI checking of the isort usage, since its easy to get the sectioning correct by hand.

If we add 4 then I tend to agree with @jakelishman that it's adding the barrier for contributors and for not-much benefit. I'm skeptical about the claimed merge-conflict resolution.

As a compromise we could turn off `only_sections` for the purposes of generating the isort-ed code to check in for this PR, since there will be a ton of churn anyway with this PR. But then leave `only_sections=True` in the config-file for the future. That way it will presumably take a while for the codebase to drift away from that sorting and if contributors happen to have tooling setup that sorts within sections they won't be adding additional churn.

**Comment by ikkoham** [comment]
Close this PR. If someone wants to continue this, please reopen.

**Pull request closed without merging**